State machine editor THO-197

PabioGF opened the pull request:
State machine editor finished and ready to merge.

I have been few days now testing possible problems that could arise and fixing them but surely there are more that I have not thought about so please, test everything you can think of. Right now the links between states when loading a state machine are not drawn but they exist, happens the same when inputing a link to a state that already has an inputed link. This is a problem that I have been trying to fix for a while now and I will try to fix it in the future, but it seems to me a minor problem right now and I have decided to make the pull request since I talked with Lukas that the state machine should be already implemented. 

To open a state machine editor you only have to click View->Engine Editor Window->State Machine Editor Engine Editor.
To test it you can do different functions, the first one is to create nodes with the right click and delete them by right clicking on one.
<img width="956" alt="Delete" src="https://github.com/user-attachments/assets/04c59c95-f66f-48f3-bc44-dd9f06344c72" />

Also, clicking on a node will open the inspector window, where information about the node will be displayed.
In this inspector you can change the name of the state. Two states can't have the same name. 
You can also modify the clip that this state has associated in a Dropbox where all the animations that the user has saved will appear. This part along the Triggers will be done in the task that joins the state machine and the animations.
Information of the associated clip is also shown, as well as a checkbox to modify if the animation is looped or not.
Finally, the transitions that this state has associated are shown, as well as parameters to modify them such as the transition duration time and the trigger that activates it, which can also be selected from a list.

<img width="711" alt="image" src="https://github.com/user-attachments/assets/e166e208-2472-487b-bdc4-642949c06095" />

Finally, you can save and load your state machines with the buttons. To save you must enter a name, and if it already exists it will overwrite the content. When loading, a list of all the state machines that the user has saved is displayed for the user to select one.
<img width="287" alt="image" src="https://github.com/user-attachments/assets/9e476456-c04d-4307-a7c3-871b1307173c" />
<img width="406" alt="image" src="https://github.com/user-attachments/assets/09806322-8903-4911-a708-82228d9d4f12" />

LukasGanter reviewed (dismissed):
The basic functionality is there and works well, good job!
Some issues i found:
- It is not possible to connect multiple transitions to the input of a node. This is necessary though for the state machine (Unless it is possible to use a connection in both directions?, but then we need to transition triggers for a single connection, one for each direction)
- When connecting a connection to the input of a node which already has an input connection (and the old connection gets deleted automatically), the connected transitions property in the state inspector is not updated and still shows the old connection as well as the new one
- Pressing the "Delete" key on the keyboard with a node selected crashes the engine

PabioGF commented:
> The basic functionality is there and works well, good job! Some issues i found:
> 
> * It is not possible to connect multiple transitions to the input of a node. This is necessary though for the state machine (Unless it is possible to use a connection in both directions?, but then we need to transition triggers for a single connection, one for each direction)
> * When connecting a connection to the input of a node which already has an input connection (and the old connection gets deleted automatically), the connected transitions property in the state inspector is not updated and still shows the old connection as well as the new one
> * Pressing the "Delete" key on the keyboard with a node selected crashes the engine

Hi, I'm aware of the problem that if you connect and input that do already has one, the previous delete, I wrote it on the PR description, however just remark that the transition is still there, is not deleted, is just that it doesn't draw on the graph for some issue with the ImNodeFlow, but the transition is not deleted. For that i think that is a minor issue that i will try to fix during the next task.

I will look now into the Delete thing and try to fix it! Thanks!

PabioGF commented:
Delete error fixed now, please try it again however, thanks!

SirKaza commented on SobrassadaEngine/FileSystem/StateMachineManager.cpp:
```diff
@@ -3,136 +3,131 @@
-#include "ResourceStateMachine.h"
+#include "MetaStateMachine.h"
+#include "ProjectModule.h"
+#include <ResourceStateMachine.h>
```
"", <> only for external std libraries

SirKaza commented on SobrassadaEngine/FileSystem/StateMachineManager.cpp:
```diff
@@ -3,136 +3,131 @@
+
+        rapidjson::Value stateMachineJSON(rapidjson::kObjectType);
+
+        UID uid              = GenerateUID();
```
const

SirKaza commented on SobrassadaEngine/FileSystem/StateMachineManager.cpp:
```diff
@@ -3,136 +3,131 @@
+        rapidjson::Value stateMachineJSON(rapidjson::kObjectType);
+
+        UID uid              = GenerateUID();
+        std::string savePath = STATEMACHINES_LIB_PATH + std::string("StateMachine") + STATEMACHINE_EXTENSION;
```
Why it is declared here and redefined after? Define it only once and use const

SirKaza commented on SobrassadaEngine/FileSystem/StateMachineManager.cpp:
```diff
@@ -3,136 +3,131 @@
+
+         // Create meta file
+        std::string assetPath = STATEMACHINES_LIB_PATH + stateName + STATEMACHINE_EXTENSION;
+        MetaStateMachine meta(stateMachineUID, assetPath);
```
Metadata is being saved in library of the engine?

SirKaza commented on SobrassadaEngine/FileSystem/StateMachineManager.cpp:
```diff
@@ -3,136 +3,131 @@
         }
-        else
+
+        std::string machinePath = STATEMACHINES_LIB_PATH + std::to_string(stateMachineUID) + STATEMACHINE_EXTENSION;
```
I don't understand why is not project library folder

SirKaza commented on SobrassadaEngine/FileSystem/StateMachineManager.h:
```diff
@@ -6,17 +6,13 @@
+    UID Save(ResourceStateMachine* stateMachine, bool override);
+    void CopyMachine(const std::string& filePath, const std::string& targetFilePath, const std::string& name, const UID sourceUID);
     ResourceStateMachine* Load(UID stateMachineUID);
+    std::vector<std::string> GetAllStateMachineNames();
```
Return const & instead of a copy

SirKaza commented on SobrassadaEngine/Modules/EditorUIModule.h:
```diff
@@ -3,7 +3,10 @@
 #include "Globals.h"
 #include "Module.h"
-#include "NodeEditor.h"
+#include <ComponentUtils.h>
```
""

SirKaza commented on SobrassadaEngine/ResourceManagement/Resources/Resource.h:
```diff
@@ -15,6 +15,7 @@ enum class ResourceType
     Animation, // 15
     Prefab,    // 16
+    StateMachine,//17
     Navmesh,   // 17
```
navmesh is 18 then

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+#include "UI/EngineEditors/Editor/StateMachineEditor.h"
```
Look include directories for Engine solution, UI/EngineEditors/Editor/ is not needed

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+    ShowLoadPopup();
+    graph->update();
+
+    for (auto& pair : graph->getNodes())
```
const

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+
+    for (auto& pair : graph->getNodes())
+    {
+        auto& node = pair.second;
```
const

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+            {
+                if (ImGui::MenuItem("Delete State"))
+                {
+                    auto stateNode = dynamic_cast<StateNode*>(node);
```
Why auto? You know this will be StateNode*

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+    ImGui::Text("State");
+    ImGui::Separator();
+
+    std::string stateName = selectedNode->GetStateName();
```
const &

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+    selectedNode = nullptr;
+    if (graph)
+    {
+        for (auto& [uid, node] : graph->getNodes())
```
const

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+{
+    for (const auto& linkWeak : graph->getLinks())
+    {
+        if (auto link = linkWeak.lock())
```
const

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+
+            if (!startPin || !endPin) continue;
+
+            auto startNode = dynamic_cast<StateNode*>(startPin->getParent());
```
const

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+
+void StateMachineEditor::ShowSavePopup()
+{
+    static bool initialized = false;
```
avoid static (bad practice)

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+void StateMachineEditor::ShowLoadPopup()
+{
+    static std::vector<std::string> stateMachineNames;
+    static int selectedIndex = -1;
```
avoid static (bad practice)

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+
+        if (ImGui::Button("Load") && selectedIndex >= 0)
+        {
+            std::string selectedName = stateMachineNames[selectedIndex];
```
this can be const & if stateMachineNames is class varible

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+                if (stateMachine)
+                {
+                    GLOG("Successfully loaded State Machine: %s", selectedName.c_str());
+                    resource = const_cast<ResourceStateMachine*>(stateMachine);
```
avoid const_cast (bad practice)

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+
+void StateMachineEditor::DeleteStateResource(StateNode& node)
+{
+    GLOG("hola");
```
hola

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+void StateMachineEditor::DeleteStateResource(StateNode& node)
+{
+    GLOG("hola");
+    std::string stateName = node.GetStateName();
```
const &

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+    }
+
+    auto state = resource->GetState(stateName);
+    clipName   = state->clipName.GetString();
```
const &

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+        }
+    }
+
+    auto state = resource->GetState(stateName);
```
remove auto and const *

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,677 @@
+    clipName   = state->clipName.GetString();
+    resource->RemoveState(state->name.GetString());
+
+    auto clip = resource->GetClip(clipName);
```
remove auto and const *

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.h:
```diff
@@ -0,0 +1,40 @@
+#include "UI/EngineEditors/EngineEditorBase.h"
+#include "ResourceManagement/Resources/ResourceStateMachine.h"
+#include "Libs/ImNodeFlow-1.2.2/include/ImNodeFlow.h"
+#include "UI/EngineEditors/Nodes/StateNode.h"
```
forward declaration

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.h:
```diff
@@ -0,0 +1,40 @@
+#include "Globals.h"
+#include "UI/EngineEditors/EngineEditorBase.h"
+#include "ResourceManagement/Resources/ResourceStateMachine.h"
+#include "Libs/ImNodeFlow-1.2.2/include/ImNodeFlow.h"
```
forward declaration

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.h:
```diff
@@ -0,0 +1,40 @@
+
+#include "Globals.h"
+#include "UI/EngineEditors/EngineEditorBase.h"
+#include "ResourceManagement/Resources/ResourceStateMachine.h"
```
forward declaration

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Nodes/StateNode.h:
```diff
@@ -0,0 +1,33 @@
+#pragma once
+#include "Libs/ImNodeFlow-1.2.2/include/ImNodeFlow.h"
```
look includes dir pls

SirKaza replied:
Libs/ImNodeFlow-1.2.2/include/ this is not needed

PabioGF commented:
Thanks for all the comments! helps to improve mi coding skills jeje

I have changed everything that i could change without the code to literally explode, that is all the comments except one or two. There was an important error in the paths of the save function, thanks you noticed it, i think it's now fixed.
Hope this is better for the PR, please check it again when you can.

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Nodes/StateNode.h:
```diff
@@ -0,0 +1,33 @@
+#pragma once
+#include "Libs/ImNodeFlow-1.2.2/include/ImNodeFlow.h"
+#include "ResourceManagement/Resources/ResourceStateMachine.h"
```
ResourceManagement/Resources/ not needed

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.h:
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "Globals.h"
+#include "UI/EngineEditors/EngineEditorBase.h"
```
UI/EngineEditors/ not needed

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.h:
```diff
@@ -0,0 +1,43 @@
+
+#include "Globals.h"
+#include "UI/EngineEditors/EngineEditorBase.h"
+#include "ImNodeFlow-1.2.2/include/ImNodeFlow.h"
```
ImNodeFlow-1.2.2/include/ not needed

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.h:
```diff
@@ -0,0 +1,43 @@
+
+
+  private:
+    UID uid;
```
Why StateMachineEditor has UID? His parent EngineEditorBase has already an UID, this also applies to any other engine editor windows that we currently have

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/StateMachineEditor.cpp:
```diff
@@ -0,0 +1,672 @@
+#include "StateMachineEditor.h"
+#include "StateNode.h"
+#include "FileSystem/StateMachineManager.h"
+#include "Application.h"
```
This includes are not formatted, try formatting everything before merge

SirKaza reviewed (changes requested):
Deleting a state when connected crashes the engine sometimes:
Exception thrown: read access violation.
ResourceStateMachine::GetState(...) returned nullptr.

Save of a state machine doesn't work. (SceneImporter::CreateDirectories is not creating directories for state machines).
Apart from that, tested everything else and works awesome.

SirKaza reviewed (approved):
Could be nice if you could fix the visual bug of relations between states, so each time it loads a state machine, you see the connections between them. Apart from that, nicely done. 🧇
